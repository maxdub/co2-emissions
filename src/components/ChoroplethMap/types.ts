type Co2Data = {
    co2: number;
    co2_per_capita: number;
    coal_co2: number;
    coal_co2_per_capita: number;
    cumulative_co2: number;
    cumulative_coal_co2: number;
    population: number;
    share_global_co2: number;
    share_global_coal_co2: number;
    share_global_cumulative_co2: number;
    share_global_cumulative_coal_co2: number;
    year: number;
};

type CountryData = {
    iso_code: string;
    data: Array<Co2Data>;
};

export type CountryCo2Data = {
    [key: string]: CountryData;
};
