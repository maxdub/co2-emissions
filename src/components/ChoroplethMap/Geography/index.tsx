import React from "react";

function Geography() {
    return <></>;
}

export default Geography;
