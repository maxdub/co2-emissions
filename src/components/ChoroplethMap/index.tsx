import React, { useEffect, useState } from "react";
import { ComposableMap, Geographies, Sphere, Graticule } from "react-simple-maps";
import { CountryCo2Data } from "./types";
import Geography from "../Geography";
import styled from "styled-components/macro";
import Slider from "../Slider";
import ReactTooltip from "react-tooltip";

const projectionConfig = {
    scale: 147,
};

function ChoroplethMap() {
    const [data, setData] = useState<CountryCo2Data>({});
    const [geo, setGeo] = useState({});
    const [loading, setLoading] = useState(false);
    const [curYear, setYear] = useState<number>(2019);
    const [content, setTooltip] = useState("");

    useEffect(() => {
        setLoading(true);
        Promise.all([
            fetch("https://raw.githubusercontent.com/owid/co2-data/master/owid-co2-data.json")
                .then((response) => response.json())
                .then((data) => setData(data)),
            fetch(
                "https://raw.githubusercontent.com/zcreativelabs/react-simple-maps/master/topojson-maps/world-110m.json",
            )
                .then((response) => response.json())
                .then((data) => setGeo(data)),
        ]).finally(() => {
            setLoading(false);
        });
    }, []);

    if (!data) {
        return null;
    }

    return (
        <>
            <ReactTooltip>{content}</ReactTooltip>
            <Slider
                curYear={curYear}
                onChange={(_evt, value) => {
                    if (Array.isArray(value)) return;
                    setYear(value);
                }}
            />
            {loading && <Loading>Loading CO2 data...</Loading>}
            <ComposableMap data-tip="World" height={500} projectionConfig={projectionConfig}>
                <Sphere fill="#F8F8F8" id="choropleth" stroke="#E4E5E6" strokeWidth={0.5} />
                <Graticule stroke="#E4E5E6" strokeWidth={0.5} />
                {Object.keys(data).length > 0 && (
                    <Geographies geography={geo}>
                        {({ geographies }) =>
                            geographies.map((geo) => (
                                <Geography
                                    key={geo.rsmKey}
                                    setTooltip={setTooltip}
                                    data={data}
                                    curYear={curYear}
                                    geo={geo}
                                />
                            ))
                        }
                    </Geographies>
                )}
            </ComposableMap>
        </>
    );
}

const Loading = styled.p`
    position: absolute;
    top: 140px;
`;

export default ChoroplethMap;
