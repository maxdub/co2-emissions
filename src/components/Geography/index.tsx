import React, { useMemo } from "react";
import { Geography as SimpleMapsGeography } from "react-simple-maps";
import { scaleLinear } from "d3-scale";
import { CountryCo2Data } from "../ChoroplethMap/types";
import { TooltipContent } from "../Tooltip";

const colorScale = scaleLinear().domain([0, 100, 500, 5000, 10000]).range([0, 150, 200, 225, 255]);

export type Props = {
    geo: any; //@@WISH ?
    data: CountryCo2Data;
    curYear: number;
    setTooltip: (content: string) => void;
};

function Geography({ geo, data, curYear, setTooltip }: Props) {
    const key: string | undefined = useMemo(
        () => Object.keys(data).find((s: string) => data[s].iso_code === geo.properties.ISO_A3),
        [data, geo],
    );

    const year = useMemo(
        () =>
            key
                ? data[key].data.find(({ year }) => {
                      return year === curYear;
                  })
                : null,
        [data, key, curYear],
    );

    return (
        <SimpleMapsGeography
            geography={geo}
            onMouseOver={(event) => {
                setTooltip(`${geo.properties.NAME} - CO2 PPM : ${year && year.co2 ? year.co2.toFixed(2) : 0}`);
            }}
            onMouseOut={() => {
                setTooltip("");
            }}
            style={{
                default: {
                    outline: "none",
                },
                hover: {
                    stroke: "#FFFFFF",
                    outline: "none",
                },
                pressed: {
                    stroke: "#FFFFFF",
                    outline: "none",
                },
            }}
            fill={`rgba(${year ? colorScale(year.co2 ? year.co2 : 0) : 0}, 98, 71, 1)`}
        />
    );
}

export default Geography;
