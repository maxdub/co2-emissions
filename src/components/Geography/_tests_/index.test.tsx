import React from "react";
import { createRenderer } from "react-test-renderer/shallow";
import { render } from "@testing-library/react";
import Geography from "../";
import { CountryCo2Data } from "../../ChoroplethMap/types";
import { Props } from "../";

const renderer = createRenderer();

const renderComponent = ({ geo, curYear, data, setTooltip }: Props) => {
    return render(<Geography setTooltip={setTooltip} geo={geo} curYear={curYear} data={data} />);
};

describe("<Geography/>", () => {
    const geo: {} = {};
    const curYear: number = 1991;
    const data: CountryCo2Data = {};
    const setTooltip = () => null;

    it("should render and match snapshot", () => {
        const output = renderComponent({ geo, curYear, setTooltip, data });
        expect(output.container.firstChild).toMatchSnapshot();
    });
});
