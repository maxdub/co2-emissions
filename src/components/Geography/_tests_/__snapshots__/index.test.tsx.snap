// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Geography/> should render and match snapshot 1`] = `
<path
  class="rsm-geography "
  fill="rgba(0, 98, 71, 1)"
  style="outline: none;"
  tabindex="0"
/>
`;
